packaging/test_plugin_unified_storage_tiering.py

```diff
@@ -776,6 +776,176 @@ def test_put_and_get(self):
+
+                    sleep(15)
+                    # stage to tier 2, look for replica in tier 0 and tier 2
+                    invoke_storage_tiering_rule()
+                    admin_session.assert_icommand('iqstat', 'STDOUT_SINGLELINE', 'irods_policy_storage_tiering')
+
+                    delay_assert_icommand(admin_session, 'ils -L ' + filename, 'STDOUT_SINGLELINE', '0 ufs0')
+                    delay_assert_icommand(admin_session, 'ils -L ' + filename, 'STDOUT_SINGLELINE', '2 ufs2')
+
+                    # test restage to tier 1
+                    admin_session.assert_icommand('iget ' + filename + ' - ', 'STDOUT_SINGLELINE', 'TESTFILE')
+
+                    sleep(15)
+                    delay_assert_icommand(admin_session, 'ils -L ' + filename, 'STDOUT_SINGLELINE', '0 ufs0')
+                    delay_assert_icommand(admin_session, 'ils -L ' + filename, 'STDOUT_SINGLELINE', '3 ufs1')
+                    delay_assert_icommand(admin_session, 'ils -L ' + filename, 'STDOUT_SINGLELINE', '2 ufs2')
+
+
+                finally:
+                    admin_session.assert_icommand('irm -f ' + filename)
+
+    def test_put_and_get_02(self):
+        with storage_tiering_configured_with_log():
+            IrodsController().restart(test_mode=True)
+            with session.make_session_for_existing_admin() as admin_session:
+                filename = 'test_put_file'
+
+                try:
+                    # make sure replicas stored on tier 1 are preserved and the minimum_restage_tier is not set
+                    admin_session.assert_icommand('imeta add -R ufs1 irods::storage_tiering::preserve_replicas true')
+                    admin_session.assert_icommand('imeta rm  -R ufs1 irods::storage_tiering::minimum_restage_tier true')
+
+                    # make sure tier 2 does not get involved
+                    admin_session.assert_icommand('imeta rm -R ufs2 irods::storage_tiering::group example_group 2')
+
+                    # create test file and upload it
+                    lib.create_local_testfile(filename)
+                    admin_session.assert_icommand('iput -R ufs0 ' + filename)
+                    admin_session.assert_icommand('ils -L ', 'STDOUT_SINGLELINE', 'rods')
+
+                    # stage to tier 1, look for both replicas
+                    sleep(6)
+                    invoke_storage_tiering_rule()
+                    admin_session.assert_icommand('iqstat', 'STDOUT_SINGLELINE', 'irods_policy_storage_tiering')
+
+                    delay_assert_icommand(admin_session, 'ils -L ' + filename, 'STDOUT_SINGLELINE', '0 ufs0')
+                    delay_assert_icommand(admin_session, 'ils -L ' + filename, 'STDOUT_SINGLELINE', '1 ufs1')
+
+                    # remove the object in tier 0
+                    admin_session.assert_icommand('itrim -N1 -n0 ' + filename, 'STDOUT_SINGLELINE', 'Number of files trimmed = 1' )
```

Comment:
This shows the problem of issue 233 I saw.
I like it because it demostrates the way we (I) want to implement the storage tiering with two tiers.
Both have preserve_replicas set to true. And we remove an object using itrim from tier 0 if space is needed using a seperate reoccuring rule.
I changed the name to: `test_preserve_replicas_works_with_restage_when_replica_only_exists_in_last_tier__issue_233`

Reply:
Ah, okay. The updated name helped me see what's going on. Thanks!